src/codegen/cfg.rs

```diff
@@ -410,6 +410,7 @@ impl BasicBlock {
 #[derive(Clone)]
 pub struct ControlFlowGraph {
     pub name: String,
+    pub original_name: String,
```

Comment:
At the moment contract functions end up in the wasm with the actual original name in Soroban. If there's a `max` function, the wasm ends up with a `max` export.

For example, the following Soroban Rust contract:
```
#![no_std]
use soroban_sdk::contractimpl;

pub struct Contract;

#[contractimpl]
impl Contract {
    pub fn max(x: u64, y: u64) -> u64 {
        x.max(y)
    }
}
```

Results in the following wasm (truncated), where the `;3;` function is exported as `max`:
```wat
(module
  (type (;0;) (func (param i64) (result i64)))
  (type (;1;) (func (param i32 i64)))
  (type (;2;) (func (param i64 i64) (result i64)))
  (type (;3;) (func))
  (import "i" "0" (func (;0;) (type 0)))
  (import "i" "_" (func (;1;) (type 0)))
  (func (;2;) (type 1) (param i32 i64)
    ...)
  (func (;3;) (type 2) (param i64 i64) (result i64)
    ...)
  (func (;4;) (type 3))
  (memory (;0;) 16)
  (global (;0;) (mut i32) (i32.const 1048576))
  (global (;1;) i32 (i32.const 1048576))
  (global (;2;) i32 (i32.const 1048576))
  (export "memory" (memory 0))
  (export "max" (func 3))
  (export "_" (func 4))
  (export "__data_end" (global 1))
  (export "__heap_base" (global 2)))
```

In this hack we modified Solang to preserve the original function name and then hacked it to use that name in the emitted functions, just so we could get it minimally working. 

There's a problem here in that Solidity supports function overloading of both regular functions and constructors, and Soroban and WASM requires a function to have a unique name. cc @graydon

Reply:
Ahh, yes I think that would work for constructors.

For function name mangling, at the moment we still need to stay under 10 characters. Maybe a truncated hash would be best. Although, it will mean other Soroban contracts would be calling functions like `.98a7df9(x, y)` instead of `.max(x, y)`.

@graydon This could be a good argument for supporting a mapping of rendered function names to exported WASM functions in the Soroban contract specification XDR. It would be a simple addition, and would allow code generators to generate nice function names like `max` while the actual WASM function would be named `98a7df9`.